go/service/saltpack.go

```diff
@@ -337,6 +354,70 @@ func (h *SaltpackHandler) SaltpackEncryptFile(ctx context.Context, arg keybase1.
 	}, nil
 }
 
+func (h *SaltpackHandler) saltpackEncryptDirectory(ctx context.Context, arg keybase1.SaltpackEncryptFileArg) (keybase1.SaltpackEncryptFileResult, error) {
+	h.G().Log.Debug("encrypting directory %s", arg.Filename)
+
+	progReporter := func(bytesComplete, bytesTotal int64) {
+		h.G().NotifyRouter.HandleSaltpackOperationProgress(ctx, keybase1.SaltpackOperationType_ENCRYPT, arg.Filename, bytesComplete, bytesTotal)
+	}
+	prog, err := newDirProgressWriter(progReporter, arg.Filename)
+	if err != nil {
+		return keybase1.SaltpackEncryptFileResult{}, err
+	}
+
+	zipch, pipeRead := zipDir(arg.Filename, prog)
+
+	outFilename, bw, err := boxFilename(arg.Filename, encryptedDirSuffix, arg.DestinationDir)
+	if err != nil {
+		return keybase1.SaltpackEncryptFileResult{}, err
+	}
+	defer bw.Close()
+
+	opts := h.encryptOptions(arg.Opts)
+	opts.Binary = true
+	opts.UseDeviceKeys = true
+
+	earg := &engine.SaltpackEncryptArg{
+		Opts:   opts,
+		Sink:   bw,
+		Source: pipeRead,
+	}
+
+	h.G().NotifyRouter.HandleSaltpackOperationStart(ctx, keybase1.SaltpackOperationType_ENCRYPT, arg.Filename)
+
+	usedSBS, assertion, err := h.frontendEncrypt(ctx, arg.SessionID, earg)
+	if err != nil {
+		h.G().Log.Debug("encrypt error, so removing %q", outFilename)
+		if clErr := bw.Close(); clErr != nil {
+			h.G().Log.Debug("error closing bw for %q: %s", outFilename, clErr)
+		}
+		if rmErr := os.Remove(outFilename); rmErr != nil {
+			h.G().Log.Debug("error removing %q: %s", outFilename, rmErr)
+		}
+		return keybase1.SaltpackEncryptFileResult{}, err
+	}
+
+	ziperr := <-zipch
```

Comment:
it's kind of funny that this can error without causing frontendEncrypt to error. Maybe there's a way to poison the reader instead.

Reply:
There is `func (*PipeWriter) CloseWithError`

Reply:
Good call.